syntax = "proto3";
package Proto;

import "Exchange.proto";
import "Request.proto";
import "Reply.proto";
import "Currency.proto";

enum InstrumentType
{
  Stock = 0;
  Future = 1;
  Option = 2;
}

enum InstrumentStatus
{
  Unknown = 0;
  Closed = 1;
  PreOpen = 2;
  OpeningAuction = 3;
  Trading = 4;
  Fuse = 5;
  Auction = 6;
  ClosingAuction = 7;
  Halt = 8;
}

enum OptionType
{
  Call = 0;
  Put = 1;
}

enum ExerciseType
{
  European = 0;
  American = 1;
}

enum SettlementType
{
  CashSettlement = 0;
  PhysicalSettlement = 1;
}

enum CommissionType
{
  Volume = 0;
  Money = 1;
}

enum RoundDirection
{
  Nearest = 0;
  Up = 1;
  Down = 2;
}

message Instrument
{
  string id = 1;
  string symbol = 2;
  string product = 3;
  Exchange exchange = 4;
  InstrumentType type = 5;
  Currency currency = 6;
  string underlying = 7;
  string hedge_underlying = 8;
  int32 lot = 9;
  double tick = 10;
  double multiplier = 11;
  double highest = 12;
  double lowest = 13;
  CommissionType commission = 14;
  double open_commission = 15;
  double close_commission = 16;
  double close_today_commission = 17;
  OptionType call_put = 18;
  string maturity = 19;
  ExerciseType exercise = 20;
  SettlementType settlement = 21;
  double strike = 22;
  InstrumentStatus status = 23;
}

message InstrumentReq
{
  RequestType type = 1;
  Exchange exchange = 2;
  repeated Instrument instruments = 3;
  string user = 4;
}

message InstrumentRep
{
  repeated Instrument instruments = 1;
  Reply result = 2;
}
